templates/fine-tune-llm_v2/cookbooks/optimize_cost/README.ipynb

```diff
@@ -0,0 +1,302 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Table of Content\n",
+    "\n",
+    "- [Overview](#overview)\n",
+    "- [How to benchmark things yourself?](#how-do-you-benchmark-yourself) \n",
+    "- [Configurations](#configurations)\n",
+    "    - [Instance type](#instance-type)\n",
+    "    - [Batch size and Context length](#batch-size-and-context-length)\n",
+    "    - [Deepspeed configs](#deepspeed-configs)\n",
+    "    - [Activation checkpointing](#activation-checkpointing)\n",
+    "    - [Gradient accumulation](#gradient-accumulation)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Overview\n",
+    "\n",
+    "Picking the correct type of GPUs and right combination of configuration is not easy to prescribe. It usually comes down to running benchmarks yourself and be able to quickly understand the trade-offs for speed and memory. There is a large collection of tutorials on the internets that can give you holistic view of how to think about performance of training LLMs, so that we do not too much into the details.\n",
+    "The following two are very good resources:\n",
+    "\n",
+    "- https://sumanthrh.com/post/distributed-and-efficient-finetuning/\n",
+    "- https://github.com/stas00/ml-engineering/blob/master/training/performance/README.md (more advanced)\n",
+    "\n",
+    "\n",
+    "When thinking about training models, there are two issues that come up in order of importance:\n",
+    "\n",
+    "- What parameters should I set to make sure I do not run out of GPU Memory (hit CUDA OOM) during training?\n",
+    "- What parameters would give me a better throughput and therefore reduce my cost?\n",
+    "\n",
+    "Now in order to fit a large model (say 70B sized model) the priority is first fitting the model. If the you cannot even train a model at batch size = 1, then trowing more hardware at it may not help. \n",
+    "\n",
+    "In this cookbook we will discuss the set of  configurations and how to change them from the default behavior to achieve a better cost. \n",
+    "Before that lets see how we can benchmark a certain configuration and measure efficiency measures like MFU."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## How do you benchmark yourself?\n",
+    "\n",
+    "For benchmarking we want to fix context length to a constant value so that it does not vary accross different batches. To do this we need to make padding strategy `max_length` (the default is `longest`). If you use `longest`, you may observe that you did not OOM for a few iteration but then OOMed in a later iteration and you might miss that during initial profiling. This is because the context length of the batch can vary at every iteration and we want to remove that as a confounding factor, so in the config YAML make sure you have the following:\n",
```

Comment:
```suggestion
    "For benchmarking we want to fix context length to a constant value so that it does not vary across different batches. To do this we need to make padding strategy `max_length` (the default is `longest`). If you use `longest`, you may observe that you did not OOM for a few iteration but then OOMed in a later iteration and you might miss that during initial profiling. This is because the context length of the batch can vary at every iteration and we want to remove that as a confounding factor, so in the config YAML make sure you have the following:\n",
```